Adding Depth Limited Search

Arafatk opened the pull request:
Adding Depth Limited search as per issue #1912 
@dschult @jfinkels please take a look.

jfinkels commented:
These functions are all nearly identical to those in `depth_first_search`. Instead of writing this code twice, maybe we can just add a `depth_limit` keyword argument to the existing depth-first search functions? @dschult did suggest having a separate module in https://github.com/networkx/networkx/issues/1912#issuecomment-169702983, but the depth limit seems relevant enough to the original DFS algorithm to apply as a keyword argument.

If we are going to have two separate modules, at least refactor the code so that you have a single underlying function, `_dfs_edges`, that can be called by both `dls_edges` and `dfs_edges`.

jfinkels commented:
Sorry, what I should have said in the second paragraph of my previous comment is have `dfs_edges` call `dls_edges` with a depth limit of positive infinity.

dschult commented:
Yes, I agree with @jfinkels.  This is so similar to ```dfs``` that it should just be implemented as an optional argument to the existing ```dfs``` routines.  We do this already for ```single_source_shortest_path``` with an argument called "cutoff". I like ```depth_limit``` here as @jfinkels suggested.

I hope you have solved the proxy issues. :)

Arafatk commented:
@dschult  I think you are right, depth should be an optional argument to the existing dfs routines.And in case if the optional argument is not provided then the function will take the depth as the number of nodes in the graph which is the maximum possible depth, This will function exactly like the original dfs.
I have two questions 
1. Should i make the changes such that the original depth_first_search supports depth_limit ?
2. Where should i keep the tests?( a new file or in the existing test_dfs file)

dschult commented:
1. How hard is it to compare the time of the original dfs code with the
code that supports depth_limit?  Using ipython and the ```%timeit``` magic
try timing each method for a single graph (maybe binomial_graph(2000,
0.3)?). If its within 20-30% then make the changes to original dfs. We
probably want it in there even if the timing slows more than that but we
might want to talk about it first. :}

2. The tests can go in the existing file.

On Tue, Jan 12, 2016 at 2:23 AM, Arafat <notifications@github.com> wrote:

> @dschult <https://github.com/dschult> I think you are right, depth should
> be an optional argument to the existing dfs routines.And in case if the
> optional argument is not provided then the function will take the depth as
> the number of nodes in the graph which is the maximum possible depth, This
> will function exactly like the original dfs.
> I have two questions
> 1. Should i make the changes such that the original depth_first_search
> supports depth_limit ?
> 2. Where should i keep the tests?( a new file or in the existing test_dfs
> file)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/networkx/networkx/pull/1928#issuecomment-170822190>.
>

Arafatk commented:
@dschult I compared time of the original dfs code with the code that supports depth_limit.
The time taken by the function with the depth_limit  was usually up by around 5-10% but sometimes even 20%.  I have attached an  txt file showing the running times for some randomly generated graphs
[time.txt](https://github.com/networkx/networkx/files/87796/time.txt)
I have added the tests for depth_limit in the existing dfs file.I have also corrected a few pep8 errors in the existing depth_first_search.py file such as missing white spaces etc.

dschult commented:
blank lines are needed at lines 5 and 19 (and maybe elsewhere). 

That type of white space is restricted by the formatting software we use for documentation (sphinx). Each doc_string starts with a single-line description of the object, a blank line and then an optional paragraph with more details. This is true even at the top of the file. Generally try to follow what is already present in terms of blank lines. 

The white space issues we often need to update are:
 -   spaces at the end of lines, and
 -   spaces after most commas and around operators/assignments.

dschult commented on networkx/algorithms/traversal/depth_first_search.py:
```diff
@@ -53,38 +51,42 @@ def dfs_edges(G, source=None):
         while stack:
-            parent,children = stack[-1]
+            parent, depth_now, children = stack[-1]
+            if not depth_now >= 1:
```
```if depth_now < 1:``` or maybe ```if depth_now == 0:```

Also, would it be more efficient to check depth before putting the child in the stack?

dschult commented on networkx/algorithms/traversal/depth_first_search.py:
```diff
@@ -128,29 +129,28 @@ def dfs_predecessors(G, source=None):
     """
-    return dict((t,s) for s,t in dfs_edges(G,source=source))
+    return dict((t, s) for s, t in dfs_edges(G, source=source,
+    depth_limit=depth_limit))
```
line continuations should be indented... usually up to the open parentheses that makes them a line continuation--but whatever makes it look good. If you have more than one line continuation the should be indented the same amount.

dschult commented on networkx/algorithms/traversal/depth_first_search.py:
```diff
@@ -162,33 +162,31 @@ def dfs_successors(G, source=None):
        Specify starting node for depth-first search and return edges in
        the component reachable from source.
-
+    depth_limit : length, optional
```
```length``` isn't a ```type```. This should probably be ```depth_limit : int, optional (default=len(G))```

Arafatk commented:
@dschult Thanks for the valuable pointers.I have made the changes.

jfinkels commented on networkx/algorithms/traversal/depth_first_search.py:
```diff
@@ -43,6 +48,7 @@ def dfs_edges(G, source=None):
     -----
     Based on http://www.ics.uci.edu/~eppstein/PADS/DFS.py
     by D. Eppstein, July 2004.
+    https://en.wikipedia.org/wiki/Depth-limited_search
```
Readers of the rendered documentation should usually never see raw URLs. Instead add a meaningful English sentence with a sensible hyperlink. For example,

    The implementation of this function is adapted from David Eppstein's
    depth-first search function in `PADS`_, with modifications to allow depth
    limits based on the Wikipedia article "`Depth-limited search`_".

    .. _PADS: http://www.ics.uci.edu/~eppstein/PADS
    .. _Depth-limited search: https://en.wikipedia.org/wiki/Depth-limited_search

For more information on docstring format, see the [Sphinx documentation][1].

[1]: www.sphinx-doc.org/en/stable/rest.html

jfinkels commented on networkx/algorithms/traversal/depth_first_search.py:
```diff
@@ -128,14 +147,15 @@ def dfs_predecessors(G, source=None):
     Based on http://www.ics.uci.edu/~eppstein/PADS/DFS.py
     by D. Eppstein, July 2004.
-
+    https://en.wikipedia.org/wiki/Depth-limited_search
```
Same comment as above about raw URLs.

jfinkels commented on networkx/algorithms/traversal/depth_first_search.py:
```diff
@@ -128,14 +147,15 @@ def dfs_predecessors(G, source=None):
     repeatedly until all components in the graph are searched.
     """
-    return dict((t,s) for s,t in dfs_edges(G,source=source))
+    return dict((t, s) for s, t in dfs_edges(G, source=source,
```
Since you are already changing this line, might as well use a dictionary comprehension instead (see #1640):

    return {t: s for s, t in dfs_edges(G, source, depth_limit)}

jfinkels commented on networkx/algorithms/traversal/depth_first_search.py:
```diff
@@ -200,18 +226,18 @@ def dfs_postorder_nodes(G,source=None):
     """
-    post=(v for u,v,d in nx.dfs_labeled_edges(G,source=source)
-          if d['dir']=='reverse')
+    post = (v for u, v, d in nx.dfs_labeled_edges(G, source=source,
```
Maybe better to do

    edges = nx.dfs_labeled_edges(...)
    return (v for u, v, d in edges if d['dir'] == 'reverse')

to avoid the line breaks altogether.

jfinkels commented on networkx/algorithms/traversal/depth_first_search.py:
```diff
@@ -239,18 +268,18 @@ def dfs_preorder_nodes(G, source=None):
     """
-    pre=(v for u,v,d in nx.dfs_labeled_edges(G,source=source)
-         if d['dir']=='forward')
+    pre = (v for u, v, d in nx.dfs_labeled_edges(G, source=source,
```
Same comment as above about splitting this into two lines.

jfinkels commented on networkx/algorithms/traversal/tests/test_dfs.py:
```diff
@@ -66,3 +66,62 @@ def test_dfs_tree_isolates(self):
         assert_equal(sorted(T.edges()),[])
+
+
+class Testdls:
```
This should be `class TestDepthLimitedSearch(object):` --- better name and better for Python 2.7 support.

jfinkels commented on networkx/algorithms/traversal/tests/test_dfs.py:
```diff
@@ -66,3 +66,62 @@ def test_dfs_tree_isolates(self):
+
+    def dls_test_preorder_nodes(self):
+        assert_equal(list(nx.dfs_preorder_nodes(self.G, source=0,
+        depth_limit=2)), [0, 1, 2])
```
Line continuation needs to be indented to match parentheses.

jfinkels commented:
This would really benefit from an example in the documentation for the modified functions.

Arafatk commented:
@jfinkels  I have made the changes.

Arafatk commented:
@dschult @jfinkels Can you suggest some changes to make, so that this pr gets fit for merging?
Thanks

dschult commented:
Would it be more efficient to put the check on depth_limit before the node goes on the stack instead of after it comes off the stack?  It seems like the stack would be much smaller and the number of times you hit a ```continue``` would be smaller too.

Arafatk commented:
@dschult You are right,the stack would be much smaller using this.I have made the changes.

jfinkels commented on networkx/algorithms/traversal/depth_first_search.py:
```diff
@@ -192,26 +245,33 @@ def dfs_postorder_nodes(G,source=None):
     """
-    post=(v for u,v,d in nx.dfs_labeled_edges(G,source=source)
-          if d['dir']=='reverse')
+    edges = nx.dfs_labeled_edges(G, source=source,depth_limit=depth_limit)
```
Missing a space after the comma; run PEP8 checks again.

jfinkels commented on networkx/algorithms/traversal/depth_first_search.py:
```diff
@@ -231,26 +294,33 @@ def dfs_preorder_nodes(G, source=None):
     >>> print(list(nx.dfs_preorder_nodes(G,0)))
+    [0, 1, 2, 3, 4]
+
+    >>> print(list(nx.dfs_preorder_nodes(G,0,2)))
```
You don't need to have `print` here, just do `list(...)`, since it is part of an interactive shell.

Arafatk commented:
@jfinkels Thanks,I have made the changes.

dschult commented on networkx/algorithms/traversal/depth_first_search.py:
```diff
@@ -192,26 +245,33 @@ def dfs_postorder_nodes(G,source=None):
-          if d['dir']=='reverse')
+    edges = nx.dfs_labeled_edges(G, source=source, depth_limit=depth_limit)
     # potential modification: chain source to end of post-ordering
     # return chain(post,[source])
```
You've removed "post" from the code so you either need to update this comment or remove it.

dschult commented:
Looks like the running time changes for dfs_edges should be put in place for dfs_labeled_edges.

chebee7i commented:
How did improving what enters the stack affect run times? Is the difference most gone now?

Arafatk commented:
@chebee7i The improving what enters the stack will not change the time complexity but it is  efficient because the stack size would be smaller.
@dschult Thanks for pointing that out.I have made the changes.

chebee7i commented:
Yes, I realize that, but I asked about run times, not time complexity. No big deal though.

dschult commented:
If you've still got the code for a time check, it'd be worth trying it with the newer version. Python performance is not obvious -- sometimes small changes make a large difference.

Arafatk commented:
@dschult  This is changed version is a bit faster than the old one -> [time.txt](https://github.com/networkx/networkx/files/94050/time.txt).

dschult commented:
I think the Depth Limited Search code is good, but it now has conflicts with the current master. @Arafatk are you able to fix those? If you don't have time I'll try and put the result into a new PR.

Arafatk commented:
@jfinkels is it okay if i send a new pr?

dschult commented:
Yes -- new PR should be fine. But be aware that you may be overwriting recent changes to the repo if you simply use the same files as from this PR.  That's a place to start, but then use the "diffs" to make sure your changes didn't erase someone else's work. 

OR 

You can update your local repository's master branch with ```git pull``` and then switch to your depth_limited_search branch and try ```git merge master```. The conflicts will be flagged in the file ```depth_first_search.py```.   You can search for ```=====``` to find where the conflicts are.  

In this case, they are all in the doc_string examples because ```G=nx.Graph();G.add_path([0,1,2,3])``` is now ```G=nx.path_graph(range(3))``` and similar.  I think it might be pretty easy to do the merge, fix those 4 examples and then ```git commit```.
